Source/Core/BoogiePL.atg

```diff
@@ -459,7 +459,10 @@ Function<.out List<Declaration>/*!*/ ds.>
      Expr definition = null;
      Expr/*!*/ tmp;
      Axiom ax;
+     bool revealed = false;
   .)
+  [ "revealed"                         (. revealed = true; .) 
```

Comment:
What is the use of "revealed"? Every time the function is in scope, its axiom is automatically revealed? Is it like equivalent to having automatic `reveal functionName` after `hide *` ?

I would have thought the other way: That whatever visibility functions would have for themselves by default could be overridden by statements in an implementation.

At least, this is the affordance we now provide in Dafny thanks to your hide statements.

Special rules like this remind me of the CSS `!important` and their 5 prioritization rules that they ended up implementing, with a weight depending on the number of qualifiers in each categories, IDs, classes, tagNames.

So are we envisioning other uses cases? I can think of
* Functions which are opaque by default, and can be revealed
* Functions which are revealed by default, and can be hidden
* Functions which are always opaque even with `reveal *`, unless we use their name explicitly
* Functions which are always revealed even with `hide *`, unless we use their name explicitly
* Function which can't be made opaque, always revealed
* Functions which can't be revealed, truly black boxes

My mind is not clear on all these use cases, so perhaps could you detail a bit more why this "revealed" is now necessary and why there is no other way?

Reply:
> Is it like equivalent to having automatic reveal functionName after hide * ?

Yes

> perhaps could you detail a bit more why this "revealed" is now necessary and why there is no other way?

Currently Dafny does not allow its users to reference definitions from the prelude, so functions from there can not be revealed, which means they should also not be hidden using `hide *`. In the future, the prelude should be a Dafny library that users can see and reference, and then we could allow hiding prelude functions using `hide *`.

> So are we envisioning other uses cases?

Indeed in the future we may want more complicated systems for hiding and revealing, but I want to wait for user feedback before designing that.